src/orion/algo/mofa/mofa.py

```diff
@@ -0,0 +1,386 @@
+            raise ValueError(f"index must be >= 1! (currently: {index})!\n")
+        if n_levels not in {2, 3, 5, 7, 11, 13, 17, 19}:
+            raise ValueError(
+                f"n_levels must be a small prime number! (currently: {n_levels})!\n"
+                f"The implementation additionally restricts it to <= 19."
+            )
+        if strength not in {1, 2}:
+            raise ValueError(f"strength must be 1 or 2 (currently: {strength})!\n")
+        if threshold <= 0 or threshold >= 1:
+            raise ValueError(
+                f"threshold must be strictly between 0 and 1! (currently: {threshold})!\n"
+            )
+
+        super().__init__(
+            space,
+            seed=seed,
+            index=index,
+            n_levels=n_levels,
+            strength=strength,
+            threshold=threshold,
+        )
+
+        self.cur_n_levels = self.n_levels
+        self.roi_space = None
+        self.completed_trials = []
+        self.duplicates = defaultdict(list)
+        self.converged = False
+        self.n_trials = None
+
+        # Initialize a dictionary for the frozen values of each hyperparameter
+        self.frozen_param_values = {}
+
+        # Generate the initial set of trials
+        self.roi_space = self.space
+        self.current_trials_params = self._generate_trials_unfrozen_params(
+            self.roi_space
+        )
+
+    def _generate_trials_unfrozen_params(self, roi_space):
+        olh_samples, self.cur_n_levels = sampler.generate_olh_samples(
+            roi_space, self.n_levels, self.strength, self.index, self.rng
+        )
+        self.n_trials = len(olh_samples)
+        olh_trials = sampler.generate_trials(olh_samples, roi_space)
+        return olh_trials
+
+    def seed_rng(self, seed):
+        """Seed the state of the random number generator.
+
+        Parameters
+        ----------
+        seed: int
+            Integer seed for the random number generator.
+
+        """
+        self.rng = np.random.RandomState(seed)
+
+    @property
+    def state_dict(self):
+        """Return a state dict that can be used to reset the state of the algorithm."""
+        state_dict = super().state_dict
+        state_dict["rng_state"] = self.rng.get_state()
+        state_dict["completed_trials"] = copy.deepcopy(self.completed_trials)
+        state_dict["duplicates"] = copy.deepcopy(self.duplicates)
+        state_dict["converged"] = self.converged
+        state_dict["roi_space"] = self.roi_space
+        state_dict["frozen_param_values"] = copy.deepcopy(self.frozen_param_values)
+        if hasattr(self, "current_trials_params"):
+            state_dict["current_trials_params"] = copy.deepcopy(
+                self.current_trials_params
+            )
+        return state_dict
+
+    def set_state(self, state_dict):
+        """Reset the state of the algorithm based on the given state_dict
+
+        :param state_dict: Dictionary representing state of an algorithm
```

Comment:
```suggestion
        Parameters
        ----------
        state_dict: dict
            Dictionary representing state of an algorithm
```

Reply:
Addressed in https://github.com/Epistimio/orion/pull/864/commits/b3f58efb4dc69ffc97b85c11b959110c69933d4e